cap_build.c

```diff
@@ -0,0 +1,68 @@
+#include <string.h>
+
+int main()
+{
+    // We create two capabilities: x, and y, which is derived from x. We note that, though we use
+    // `cheri_address_set()`, it doesn't actually set (i.e. _change_) the address of its arguments,
+    // but construct a new capability with the address we pass.
+    //
+    // cheri_address_set(void* __capability, uintptr_t) -> void *__capability
+    void *__capability x = cheri_ddc_get();
+    uintptr_t y_ptr = (uintptr_t) malloc(20);
+    void *__capability y = cheri_address_set(x, y_ptr);
+
+    printf("\n\tThe second capability printed below is created from the first using");
+    printf(
+        "\n\t`cheri_address_set()`. Notice that `cheri_address_set()` created a new capability,");
+    printf("\n\trather than setting the address on the first.");
+    printf("\n\n");
+    pp_cap(x);
+    pp_cap(y);
+    printf("\n");
+
+    // We create an (invalid!) capability by writing it directly.
+    void *my_cap_mem = malloc(16);
+    long cap;
+    cap = 0xdc5f400000030005;
+    memcpy(my_cap_mem + 8, &cap, 8);
+    cap = 0x00000000deadbeef;
+    memcpy(my_cap_mem + 0, &cap, 8);
+    void *__capability my_cap = (void *__capability) *(void *__capability *) my_cap_mem;
+
+    // We build a new capability using `cheri_cap_build()`. Note that this capability is _valid_,
+    // because it's derived from `x`, which is valid. The address comes from `my_cap`, which isn't
+    // valid, but we can still create a valid capability beecause `cheri_cap_build` just reads it's
+    // address bits and constructs a capability.
+
+    // cheri_cap_build(void *__capability, uintcap_t) -> void *__capability
+    void *__capability valid_cap = cheri_cap_build(x, (uintcap_t) my_cap);
+
+    printf("\n\tThe first capability printed below is created by writing its structure into");
+    printf("\n\tmemory; it's not valid, it's just a bunch of bits. The second is an equivalent");
+    printf("\n\tcapability, which is constructed using `cheri_cap_build(). Notice that the");
+    printf("\n\tsecond one is valid; that's because it's derived using a valid capability");
+    printf("\n\t(the variable `x`, which contains a copy of the DDC).");
+    printf("\n\n");
+    pp_cap(my_cap);
+    pp_cap(valid_cap);
+    printf("\n");
+
+    printf("\n\tThese functions do similar things, but they're different.");
+    printf("\n\t`cheri_address_set()` takes a cap and a ptr, and returns a new cap with the");
+    printf("\n\taddress set to the ptr passed in, derived from the cap passed in.");
+    printf("\n\t`cheri_cap_build()` takes a cap and a uintcap_t, and returns a new cap with the");
+    printf("\n\taddress stored in the uintcap_t.");
+    printf("\n");
+    printf("\n\tWhile these are very similar, `cheri_cap_build()` is intended to be used when");
+    printf("\n\tyou're creating capabilities from \"raw data\". In general, if you're unsure");
+    printf("\n\twhich to use, it's likely `cheri_address_set()`.");
+    printf("\n\n");
```

Comment:
I would either drop this whole discussion (or sprinkle it in comments appropriately), or have an example which emphasises the difference (which the rest of this example might do).

Reply:
I reworked it (with updates following @djlowther's explanation of `cheri_cap_build` from today — thanks!) and moved the relevant parts to the top of the example as an explanation of what the example is trying to demonstrate / accomplish and as a brief intro for anybody using it to learn. WDYT of the current incarnation? (Fixed in [a1b2ba1](https://github.com/capablevms/cheri-examples/pull/92/commits/a1b2ba1cc12effdf2fa5b0bce7d8ee3bf7b6ddd8))